src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc64.Vectorized.cs

```diff
@@ -0,0 +1,159 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.InteropServices;
+
+namespace System.IO.Hashing
+{
+    public partial class Crc64
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector128<ulong> LoadFromSource(ref byte source, nuint elementOffset)
+        {
+            Vector128<byte> vector = Vector128.LoadUnsafe(ref source, elementOffset);
+
+            if (BitConverter.IsLittleEndian)
+            {
+                // Reverse the byte order.
+
+                // SSSE3 is required to get PSHUFB acceleration for Vector128.Shuffle on x86/x64.
+                // However, the gains from vectorizing the rest of the operations seem to to be
+                // greater than the added cost of emulating the shuffle, so we don't require SSSE3 support.
+                vector = Vector128.Shuffle(vector,
+                    Vector128.Create((byte)0x0F, 0x0E, 0x0D, 0x0C, 0x0B, 0x0A, 0x09, 0x08, 0x07, 0x06, 0x05, 0x04, 0x03,
+                        0x02, 0x01, 0x00));
+            }
+
+            return vector.AsUInt64();
+        }
+
+        // All of these checks except the length check are elided by JIT, so the JITted implementation
+        // will be either a return false or a length check against a constant. This means this method
+        // should be inlined into the caller.
+        private static bool CanBeVectorized(ReadOnlySpan<byte> source) => VectorHelper.IsSupported && source.Length >= Vector128<byte>.Count;
+
+        // Processes the bytes in source in 128 byte chunks using intrinsics, followed by processing 16
+        // byte chunks, and then processing remaining bytes individually. Requires at least 16 bytes of data.
+        // Requires little endian byte order and support for PCLMULQDQ intrinsics on Intel architecture
+        // or AES and AdvSimd intrinsics on ARM architecture. Based on the algorithm put forth in the Intel paper
+        // "Fast CRC Computation for Generic Polynomials Using PCLMULQDQ Instruction" in December, 2009 and the
+        // Intel reference implementation.
+        // https://github.com/intel/isa-l/blob/33a2d9484595c2d6516c920ce39a694c144ddf69/crc/crc64_ecma_norm_by8.asm
+        private static ulong UpdateVectorized(ulong crc, ReadOnlySpan<byte> source)
+        {
+            Debug.Assert(CanBeVectorized(source), "source cannot be vectorized.");
+
+            // Work with a reference to where we're at in the ReadOnlySpan and a local length
+            // to avoid extraneous range checks.
+            ref byte srcRef = ref MemoryMarshal.GetReference(source);
```

Comment:
Personally I would prefer to store a reference to `ulong` rather than `byte` and in every loop iteration update the index rather than the reference, but since similar pattern was used in #83321 and approved by people more knowledgeable in this area, so I won't suggest it.

```diff
- ref byte srcRef = ref MemoryMarshal.GetReference(source);
+ ref ulong srcRef = ref Unsafe.As<byte, ulong>(ref MemoryMarshal.GetReference(source));
```